### PR TITLE
Allow spliting per year

### PR DESCRIPTION
**Pull request by SamDepardieu**
✨ Add a new parameter to the init method of the gov spider, letting us choose the year to from when to scrape the publications

### CODE REVIEW

File: wsf_scraping/spiders/gov_spider.py [hunk 1]
```diff
@@ -10,20 +11,47 @@ class GovSpider(BaseSpider):
         'JOBDIR': 'crawls/gov_uk'
     }
 
+    def __init__(self, *args, **kwargs):
+        year_before = kwargs.get('year_before', False)
+        year_after = kwargs.get('year_after', False)
+        id = kwargs.get('uuid', '')
+
+        self.uuid = id
+        if year_before:
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: How do you feel about ternary operators?
`self.year_before = year_before if year_before else ''`

File: wsf_scraping/spiders/gov_spider.py [hunk 2]
```diff
@@ -10,20 +11,47 @@ class GovSpider(BaseSpider):
...
-            self.logger.info('Initial url: %s', url)
-            yield scrapy.Request(
-                url=url,
-                callback=self.parse,
-                errback=self.on_error,
-                dont_filter=True,
-            )
+        url = 'https://www.gov.uk/government/publications'
+
+        if self.year_before or self.year_after:
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: double checking here: what happens if only one of `year_before` or `year_after` is assigned? should this be an `and` instead?

File: wsf_scraping/spiders/gov_spider.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 from scrapy.http import Request
 from .base_spider import BaseSpider
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: is `.base_spider` local? if so:

>>>
Imports should be grouped in the following order:

    Standard library imports.
    Related third party imports.
    Local application/library specific imports.

You should put a blank line between each group of imports.

Comment (jdkram) [reply wsf_scraping/spiders/gov_spider.py]: pep8 pedantry 🤓

File: wsf_scraping/spiders/gov_spider.py [hunk 4]
```diff
@@ -10,20 +11,47 @@ class GovSpider(BaseSpider):
         'JOBDIR': 'crawls/gov_uk'
     }
 
+    def __init__(self, *args, **kwargs):
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: sorry, this isn't feedback, just curiosity- why do we need the `*args` here?

File: wsf_scraping/spiders/gov_spider.py [hunk 5]
```diff
@@ -10,20 +11,47 @@ class GovSpider(BaseSpider):
...
+            }
+            query_params = urlencode(query_dict)
+            url = url + '?' + query_params
+
+        self.logger.info('Initial url: %s', url)
+        yield scrapy.Request(
+            url=url,
+            callback=self.parse,
+            errback=self.on_error,
+            dont_filter=True,
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: `filter=False`?

Comment (jdkram) [reply wsf_scraping/spiders/gov_spider.py]: 🤷‍♂️

File: wsf_scraping/spiders/gov_spider.py (hunk 3 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: Well spotted 👍

File: wsf_scraping/spiders/gov_spider.py (hunk 4 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: It's a Django die hard habit. We don't need it at all, but it's default on Django, and so it is in my brain

File: wsf_scraping/spiders/gov_spider.py (hunk 1 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: Hum, I guess it can't hurt

File: wsf_scraping/spiders/gov_spider.py (hunk 4 above)
Comment (jdkram) [reply wsf_scraping/spiders/gov_spider.py]: 👍

File: wsf_scraping/spiders/gov_spider.py (hunk 1 above)
Comment (jdkram) [reply wsf_scraping/spiders/gov_spider.py]: purely a style thing, could make a good case arguing against readability of ternary

File: wsf_scraping/spiders/gov_spider.py (hunk 2 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: No it's perfectly normal here. If only one of these is specified, it returns everything before x or everything after x

File: wsf_scraping/spiders/gov_spider.py (hunk 5 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: Scrapy internal 🤷‍♂️

File: wsf_scraping/spiders/gov_spider.py [hunk 6]
```diff
@@ -10,20 +11,47 @@ class GovSpider(BaseSpider):
         'JOBDIR': 'crawls/gov_uk'
     }
 
+    def __init__(self, *args, **kwargs):
+        year_before = kwargs.get('year_before', False)
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: oops, sorry, one more thing - is it worth clarifying in the whether `year_before` etc are inclusive? i.e. if I run it with args 2013 - 2015, is that all of 2013, 2014, 2015? (yes, on reading the code)

File: wsf_scraping/spiders/gov_spider.py (hunk 5 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: More seriously, Scrapy default behaviour is to filter everything, so it's a change to NOT filter things, and this is probably why the meta is `dont_filter`

File: wsf_scraping/spiders/gov_spider.py (hunk 1 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: In this case it's cleaner I think 👍

File: wsf_scraping/spiders/gov_spider.py (hunk 6 above)
Comment (SamDepardieu) [reply wsf_scraping/spiders/gov_spider.py]: I'll add some comments 👍

File: wsf_scraping/spiders/gov_spider.py [hunk 7]
```diff
@@ -11,21 +13,14 @@ class GovSpider(BaseSpider):
...
-        if year_before:
-            self.year_before = year_before
-        else:
-            self.year_before = ''
-
-        if year_after:
-            self.year_after = year_after
-        else:
-            self.year_after = ''
+        self.year_before = year_before if year_before else ''
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: okay yeah this does look super clean

**Comment by SamDepardieu** [comment]
@jdkram Ready for another review 👌

File: wsf_scraping/spiders/gov_spider.py [hunk 8]
```diff
@@ -40,7 +35,7 @@ def start_requests(self):
                 'official_document_status': 'all',
                 'world_locations[]': 'all',
                 'from_date': '01/01/{}'.format(self.year_after),
-                'to_date': '31/12/{}'.format(self.year_before)
+                'to_date': '31/12/{}'.format(int(self.year_before) - 1)
```
Comment (jdkram) [inline wsf_scraping/spiders/gov_spider.py]: 👍

File: wsf_scraping/spiders/gov_spider.py (hunk 5 above)
Comment (jdkram) [reply wsf_scraping/spiders/gov_spider.py]: 👍

File: wsf_scraping/spiders/gov_spider.py (hunk 2 above)
Comment (jdkram) [reply wsf_scraping/spiders/gov_spider.py]: 👍

**Comment by jdkram** [comment]
🎉 🎉 🎉

**Pull request merged**